### PR TITLE
add JustifiedHeader::new(header,justification,authorities) constructor

### PR DESCRIPTION
**Pull request by icodezjb**
add JustifiedHeader::new(header,justification,authorities) constructor for external crate

### CODE REVIEW

File: substrate/client/src/client.rs [hunk 1]
```diff
@@ -154,6 +154,14 @@ pub struct JustifiedHeader<Block: BlockT> {
 }
 
 impl<Block: BlockT> JustifiedHeader<Block> {
+	/// Creates new JustifiedHeader with given header, justification and authorities.
+	pub fn new(header: <Block as BlockT>::Header, justification: ::bft::Justification<Block::Hash>, authorities: Vec<AuthorityId>,) -> Self {
+		JustifiedHeader{
```
Comment (gavofyork) [inline substrate/client/src/client.rs]: house style puts a space between the name of the struct and the `{`, particularly when it's not inline.

**Comment by gavofyork** [comment]
minor style nit. any particular reason for introducing this?

**Comment by rphmeier** [comment]
This breaks the type-safety we have around `JustifiedHeader`, that it is only constructible with a valid justification. The constructor should check the justification and return a `Result` to maintain this property.

**Comment by icodezjb** [comment]
@gavofyork yeah, my initial intention was to build a  substrate client with file-based backend which is not in-mem and producer blocks , so i create a backend by Backend::new in substrate/client/db, and then create a client which has pub fn check_justification and  import_block.
 in pub fn check_justification:

 ```let authorities = self.authorities_at(&BlockId::Hash(parent_hash))?;```

this will be error in my backend which was created by substrate/client/db, because ``` authorities_at ``` will call the ``` cache ``` and the ``` cache ``` return "None" in substrate/client/db/lib.rs

 ``` 
impl<Block: BlockT> client::blockchain::Backend<Block> for BlockchainDb<Block> {
     ...
      fn cache(&self) -> Option<&client::blockchain::Cache<Block>> {
		None
	}
} 
``` 

So , if I want to achieve my purpose with Client::import_block,there are two ways ,
- Construct a JustifiedHeader and send it to Client::import_block
- Improve the association function cache () in substrate/client/db/lib.rs and call Client::check_justification and Client::import_block

To the second way, i am not sure how improve it, So I select the first way.

File: substrate/client/src/client.rs (hunk 1 above)
Comment (icodezjb) [reply substrate/client/src/client.rs]: improved

File: substrate/client/src/client.rs [hunk 2]
```diff
@@ -154,6 +154,24 @@ pub struct JustifiedHeader<Block: BlockT> {
 }
 
 impl<Block: BlockT> JustifiedHeader<Block> {
+	/// Creates new JustifiedHeader with given header, unchecked justification and authorities.
+	pub fn new(header: <Block as BlockT>::Header,
+			   justification: ::bft::UncheckedJustification<Block::Hash>,
+			   authorities: Vec<AuthorityId>,
+	) -> error::Result<Self> {
+		let parent_hash = header.parent_hash().clone();
+		let just = ::bft::check_justification::<Block>(&authorities[..], parent_hash, justification)
```
Comment (dvdplm) [inline substrate/client/src/client.rs]: Perhaps `just` can be named `justification` so line 169 becomes

```rust
Ok(JustifiedHeader {
  header,
  justification,
  authorities
})
```

Comment (dvdplm) [reply substrate/client/src/client.rs]: Sure, why not? :)

Comment (icodezjb) [reply substrate/client/src/client.rs]: done

**Comment by svyatonik** [comment]
@rphmeier It isn't an issue - just an optimization. When there's no cache, `Client` simply uses 'normal' `runtime::authorities` execution: https://github.com/paritytech/substrate/blob/master/substrate/client/src/client.rs#L237

**Comment by icodezjb** [comment]
Just update substrate to last version, my code is well work! @rphmeier @svyatonik

**Pull request closed without merging**